inclavared/app/src/main.rs

```diff
@@ -116,39 +116,56 @@ fn handle_client(sockfd: RawFd, upstream: &Option<String>,
     let n = tls.receive(&mut buffer).unwrap();
     info!("req: {}", String::from_utf8((&buffer[..n]).to_vec()).unwrap());
 
-    if let Some(upstream) = upstream {
-        /* fetch enclave information from upstream */
-        let n = client_fetch(&upstream, &mut buffer,
-                    tls_type, crypto, attester, verifier, mutual, enclavefile);
-        info!("message length from upstream: {}", n);
-
-        /* TODO: shit code */
-        let resp = if n > 64 {
-            let mrenclave = &buffer[0..32];
-            let mrsigner = &buffer[32..64];
-            let message = String::from_utf8((&buffer[64..n]).to_vec()).unwrap();
-            info!("message from upstream: {}", message);
-
-            let resp = json!({
-                "id": "123456",
-                "msgtype": "ENCLAVEINFO",
-                "version": 1,
-                "mrenclave": hex::encode(mrenclave),
-                "mrsigner": hex::encode(mrsigner),
-                "message": message
-            });
-            resp.to_string()
-        } else if n > 0 {
-            String::from_utf8((&buffer[..n]).to_vec()).unwrap()
-        } else {
-            String::from("reply from inclavared!\n")
-        };
-        info!("resp: {}", resp);
+    let req: Value = match serde_json::from_slice(&buffer[..n]) {
+        Ok(r) => r,
+        Err(e) => {
+            error!("json::from_slice() failed, {}", e);
+
+            /* XXX TODO: tls must be droped before enclave */
+            drop(tls);
+            enclave.destroy();
+            return;
+        }
+    };
 
-        /* response reply */
-        tls.transmit(resp.as_bytes()).unwrap();
+    if req["type"] == "GETENCLAVEINFO" {
+        if let Some(upstream) = upstream {
+            /* fetch enclave information from upstream */
+            let n = client_fetch(&upstream, &mut buffer, tls_type, crypto,
+                        attester, verifier, mutual, enclavefile);
+            info!("message length from upstream: {}", n);
+
+            /* TODO: shit code */
+            let resp = if n > 64 {
+                let mrenclave = &buffer[0..32];
+                let mrsigner = &buffer[32..64];
+                let message = String::from_utf8((&buffer[64..n]).to_vec()).unwrap();
+                info!("message from upstream: {}", message);
+
+                let resp = json!({
+                    "id": "123456",
+                    "msgtype": "ENCLAVEINFO",
+                    "version": 1,
+                    "mrenclave": hex::encode(mrenclave),
+                    "mrsigner": hex::encode(mrsigner),
+                    "message": message
+                });
+                resp.to_string()
+            } else if n > 0 {
```

Comment:
Which condition the resp length will be less than 64 but greater than 0?

Reply:
Can we combine this condition with the below 'else'?

Reply:
A server that does not support mrsigner and mrenclave, only exists in the test state, such as enclave-tls-client